src/codecs/webp/extended.rs

```diff
@@ -239,22 +241,29 @@ impl ExtendedImage {
         anim_image: &AnimatedFrame,
         background_color: Rgba<u8>,
     ) -> Option<ImageResult<Frame>> {
-        let mut buffer = vec![0; (anim_image.width * anim_image.height * 4) as usize];
+        let mut buffer = vec![0; anim_image.image.get_buf_size()];
         anim_image.image.fill_buf(&mut buffer);
+        let color_type = anim_image.image.color_type();
+        let has_alpha = match color_type {
+            ColorType::Rgba8 => true,
+            ColorType::Rgb8 => false,
+            _ => unreachable!(),
+        };
+        let pixel_len = u32::from(color_type.bytes_per_pixel());
 
         for x in 0..anim_image.width {
             for y in 0..anim_image.height {
                 let canvas_index: (u32, u32) = (x + anim_image.offset_x, y + anim_image.offset_y);
-                let index: usize = (y * 4 * anim_image.width + x * 4).try_into().unwrap();
-                canvas[canvas_index] = if anim_image.use_alpha_blending {
+                let index: usize = ((y * anim_image.width + x) * pixel_len).try_into().unwrap();
+                canvas[canvas_index] = if anim_image.use_alpha_blending && has_alpha {
                     let buffer: [u8; 4] = buffer[index..][..4].try_into().unwrap();
                     ExtendedImage::do_alpha_blending(buffer, canvas[canvas_index])
                 } else {
                     Rgba([
                         buffer[index],
                         buffer[index + 1],
                         buffer[index + 2],
-                        buffer[index + 3],
+                        if has_alpha { buffer[index + 3] } else { 255 },
```

Comment:
I referenced this section of the WebP reference: https://developers.google.com/speed/webp/docs/riff_container#assembling_the_canvas_from_frames

It has this block in the pseudocode:

```python
      if subchunk.tag == "ALPH":
        assert alpha subchunks not found in 'Frame Data' earlier
        frame_params.alpha = alpha_data
```

I believe this assertion explicitly excludes the case that this code I added is meant to handle: the claim that the image as a whole has alpha but in reality the frames are RGB. As such, any behavior we implement in this case is a "best effort" correction.

What do you think?

Reply:
That's within a frame, i.e. each animation frame should have at most one alpha subchunk. This method is about composing frames, i.e. the loop `until eof or non-ANMF chunk`. It seems that the choice of fully opaque would be a reasonable interpretation of what's given. Feel free to reference the section (with regards to the not-specified assignment of `frame_param.alpha` when no `ALPH` occurs) in a comment.

---

To be quite frank, that ""specification"" is _extremely_ disappointing and anything but specific. What kind of semantics am I to ascribe to the pseudo-code-hogwash? Garbage collection? Python?

The critical part would indeed be `frame_params.alpha` but does it get reset? Should a default of fully opaque be used? What does `render frame […]  using blending method` mean if there is no alpha since blending is specified as _always_ consuming an `.A`.

> clear canvas to ANIM.background_color or application defined color

With application defined color, if no background color is given? Or are we always free to choose?

I've seen better undergrad work.

Reply:
Yep, the reference does not provide much to work with.